src/wallet/rpcwallet.cpp

```diff
@@ -3355,41 +3371,90 @@ UniValue z_listreceivedbyaddress(const UniValue& params, bool fHelp)
     }
 
     UniValue result(UniValue::VARR);
-    std::vector<SproutNoteEntry> sproutEntries;
-    std::vector<SaplingNoteEntry> saplingEntries;
-    pwalletMain->GetFilteredNotes(sproutEntries, saplingEntries, fromaddress, nMinDepth, false, false);
 
     std::set<std::pair<PaymentAddress, uint256>> nullifierSet;
     auto hasSpendingKey = boost::apply_visitor(HaveSpendingKeyForPaymentAddress(pwalletMain), zaddr);
     if (hasSpendingKey) {
         nullifierSet = pwalletMain->GetNullifiersForAddresses({zaddr});
     }
 
+    NotesFilters filters;
+    filters.minDepth = nMinDepth;
+    filters.ignoreSpent = false;
+    filters.requireSpendingKey = false;
+    filters.cwallet = pwalletMain;
+
     if (boost::get<libzcash::SproutPaymentAddress>(&zaddr) != nullptr) {
-        for (SproutNoteEntry & entry : sproutEntries) {
+        auto pa = *boost::get<libzcash::SproutPaymentAddress>(&zaddr);
+        auto itr_sprout = pwalletMain->GetNotesByType(pa.GetHash(), NoteType::sprout, timestamp, offset, limit);
+        auto itr = itr_sprout.begin();
+
+        while (itr != itr_sprout.end()) {
+            auto wtx = pwalletMain->mapWallet[itr->hash];
+            filters.wtx = &wtx;
+            filters.nd_sprout = pwalletMain->mapWallet[itr->hash].mapSproutNoteData[*itr->jsop];
+            JSOutPoint jsop = *itr->jsop;
+            uint64_t ts = itr->timestamp;
+            uint256 txid = itr->hash;
+            filters.jsop = jsop;
+            filters.pa_sprout = pa;
+
+            ++itr;
+
+            if (filters.Common() || filters.Sprout())
```

Comment:
I've just realized a new problem. We are taking an iterator with already applied limit in `GetNotesByType` and on this result we are filtering out some notes. So if we had limit=10, but 3 notes don't pass `Common()` or `Sprout()` filter, the result will only have `7` rows. In marginal scenario it could have return empty response, even though there are notes further down the iterator which would be returned. 
This would be confusing for the client that wants to get all the notes (in chunks) because natural assumption is to stop iterating when you receive less rows them `limit`.
Offset will also not work well, because it will have "gaps" due to skipped notes. 

I think the filtering part needs to be included inside the iterator.
So unfortunately if there is need to filter the notes, I think you cannot use handy functions like `boost::advance` to skip, but go one by one and apply the filter.

Reply:
Unfortunately you are right, i didn't saw it yesterday. I will try to fix.